cclient: JSON serializer.

oopsmonk opened the pull request:
Implementing JSON serializer #50 

TODO: 
* get/add/remove neighbors, getTrytes, attachToTangle, wereAddressesSpentFrom
* unit tests

# Test Plan:
CI pass

Commit:
cclient: Implementing JSON deserialization.

tsvisabo reviewed (commented):
How about a function that does the most basic parsing and checks for error:
 `int parse_obj(const char* const obj,cJSON* json_obj) )` ?

tsvisabo reviewed (commented):
return CCLIENT_UNIMPLEMENTED when not implemented

tsvisabo reviewed (commented):
consider adding a function that populates an array from cJSON object (reduce duplicated code)

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -4,6 +4,20 @@
 #include "cJSON.h"
 #include "json_serializer.h"
 
+int address_to_flex_trit(address_t addr, const char* trytes) {
```
this method looks odd. can't we just adjust the address_t struct to contain flex_trits direclty?

oopsmonk replied:
address_t is a point of trit_array_t, regard with memory model we need to store extra info, say num_trits and num_bytes. 
@ifullgaz and @tsvisabo, do you have any suggestions about converting flex_trits from a string(trytes)? 
actually for deserialization here should check the num_trits and num_bytes that are assigned by the user.

oopsmonk replied:
will use `flex_trit_t` instead of `trit_array_t`

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -33,9 +85,53 @@ void json_get_balances_serialize_request(const serializer_t* const s,
+
+  int int_cnt = 0;
+  cJSON* current_obj = NULL;
+  cJSON_ArrayForEach(current_obj, json_item) {
```
how do you know that `balances.array` is large enough?

oopsmonk replied:
I don't know the number from the response, the user should allocate a sensible size to store the `balances.array`, if `balances.array` from the response is more then `out->balances.size` which is allocated by a user, only keep the number of `out->balances.size`.

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -52,9 +148,44 @@ void json_get_inclusion_state_serialize_request(
+  cJSON* current_obj = NULL;
+  cJSON_ArrayForEach(current_obj, json_item) {
+    if (current_obj->string != NULL) {
+      out->bitmap.array[int_cnt] = strcmp("true", current_obj->string) ? 1 : 0;
```
how do you know that `bitmap.array` is large enough? 
(also, why `bitmap`?)

oopsmonk replied:
I'll fix it, the size of `bitmap.array` should refer to the set of tips and transactions in the request.
and rename to `states`

oopsmonk commented:
Hi @tsvisabo,  parsing json objects are handled through cJSON, the rest of code is for error handling.
if we have parse_obj, like your suggestion, there will be parse_int, parse_chat/string, and parse_array_int...
but we still have to deal with errors. something like
```
if (! parse_obj("hashes", json_obj, buff))
  cJSON_Delete(json_obj);
  return error_code
```

Commit:
cclient: use flex_trit_t instead of trit_array_t

tsvisabo commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -352,11 +370,19 @@ int json_get_tips_deserialize_response(const serializer_t* const s,
+            strlen(current_obj->string), NUM_TRYTES_ADDRESS);
+        if (ret_trytes == 0) {
+          cJSON_Delete(json_obj);
+          return EXIT_FAILURE;
```
Should return `retcode_t `

oopsmonk replied:
sure, thanks.

Commit:
cclient: using common/errors.h as error codes.

Commit:
cclient: json deserialization, error code update.

addNeighbors, removeNeighbors, getTrytes, attachToTangle,
wereAddressesSpentFrom

tsvisabo commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -455,3 +508,262 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+  }
+
+  return RC_OK;
+  return RC_CCLIENT_NOT_IMPLEMENTED;
```
remove

tsvisabo commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -455,3 +508,262 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+// addNeighbors
+size_t json_add_neighbors_serialize_request_get_size(
+    const serializer_t* const s) {
+  return 0;
```
not implemented

tsvisabo commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -455,3 +508,262 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+}
+void json_add_neighbors_serialize_request(const serializer_t* const s,
+                                          char* out) {}
+
```
not implemented

tsvisabo commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -455,3 +508,262 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+
+  json_value = cJSON_GetObjectItemCaseSensitive(json_obj, "error");
+  if (cJSON_IsString(json_value) && (json_value->string != NULL)) {
+    cJSON_Delete(json_obj);
```
Can you please add error log on each error?

tsvisabo commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -455,3 +508,262 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+
+size_t json_add_neighbors_deserialize_response_get_size(
+    const serializer_t* const s, const char* const toDeserialize) {
+  return 0;
```
not implemented

oopsmonk replied:
as we discussed, may remove this function

tsvisabo commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -455,3 +508,262 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+// removeNeighbors
+size_t json_remove_neighbors_serialize_request_get_size(
+    const serializer_t* const s) {
+  return 0;
```
not implemented

tsvisabo commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -455,3 +508,262 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+}
+
+size_t json_remove_neighbors_deserialize_response_get_size(
+    const serializer_t* const s, const char* const toDeserialize) {
```
same

th0br0 commented on common/errors.h:
```diff
@@ -89,6 +90,10 @@ enum retcode_t {
+
+  RC_CCLIENT_NOT_IMPLEMENTED =
+      0x01 | RC_MODULE_CCLIENT_JSON | RC_SEVERITY_MAJOR,
+  RC_CCLIENT_JSON_PARSE = 0x02 | RC_MODULE_CCLIENT_JSON | RC_SEVERITY_MAJOR,
```
could you add a comment explaining the error?

th0br0 commented on cclient/response/addresses_spent_from.h:
```diff
@@ -0,0 +1,20 @@
+
+typedef struct {
+  int_array states;
+} addresses_spent_from_res_t;
```
the actual api call is called `wereAddressesSpentFrom`

```
                case "wereAddressesSpentFrom": {
                    final List<String> addresses = getParameterAsList(request,"addresses", HASH_SIZE);
                    return wereAddressesSpentFromStatement(addresses);
                }
```

Could you adjust the naming here & of the files?

oopsmonk replied:
done

th0br0 commented on cclient/response/get_balances.h:
```diff
@@ -12,7 +12,7 @@ extern "C" {
   int_array balances;
   int milestoneIndex;
-  const address_t milestone;
+  const flex_trit_p milestone;
```
`flex_trit_p` is a pointer, not sure if marking it as const is useful here? (as the data itself isn't const-protected)

th0br0 commented on cclient/response/get_neighbors.h:
```diff
@@ -9,9 +9,9 @@ extern "C" {
 typedef struct {
   /**
-   * Address of your peer.
+   * IP Address of your peer.
```
it's more like `uri` because it contains either [IP:port] or [DOMAIN:port]

th0br0 reviewed (changes requested):
What about tests?

Commit:
cclient: rename addresses_spent_from to were_addresses_spent_from

Fixed parsing wrong filed in json string object.
Added json_ints_to_int_array, json_booleans_to_int_array,
and json_array_to_flex_trirts_array.

oopsmonk commented:
TODO:
implement serialization functions 
unit tests will implemented in a new PR

tsvisabo committed:
Merge branch 'develop' into json_serialization

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -124,3 +502,191 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+    return RC_CCLIENT_JSON_PARSE;
+  }
+
+  retcode_t ret = json_booleans_to_int_array(json_obj, "states", &out->states);
```
missing a `cJSON_Delete` here

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -124,3 +502,191 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+
+  retcode_t ret = json_array_to_flex_trirts_array(
+      json_obj, "trytes", &out->trytes, NUM_TRITS_SERIALIZED_TRANSACTION,
+      NUM_TRYTES_SERIALIZED_TRANSACTION);
```
missing a `cJSON_Delete`

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -124,3 +502,191 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+
+  retcode_t ret = json_array_to_flex_trirts_array(
+      json_obj, "trytes", &out->trytes, NUM_TRITS_SERIALIZED_TRANSACTION,
+      NUM_TRYTES_SERIALIZED_TRANSACTION);
```
missing a `cJSON_Delete`

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -124,3 +502,191 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+    return RC_CCLIENT_JSON_PARSE;
+  }
+
+  *out = json_value->valueint;
```
missing a cJSON_Delete

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -124,3 +502,191 @@ size_t json_get_transactions_to_approve_deserialize_response_get_size(
+    return RC_CCLIENT_JSON_PARSE;
+  }
+
+  *out = json_value->valueint;
```
missing a cJSON_Delete

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -111,9 +450,48 @@ size_t json_get_tips_deserialize_response_get_size(
+      cJSON_Delete(json_obj);
+      return RC_CCLIENT_JSON_PARSE;
+    }
+  }
```
missing a cJSON_Delete

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -88,19 +298,148 @@ void json_get_node_info_serialize_request(const serializer_t* const s,
+  retcode_t ret = json_array_to_flex_trirts_array(
+      json_obj, "hashes", &out->tips, NUM_TRITS_ADDRESS, NUM_TRYTES_ADDRESS);
+  if (ret != RC_OK) return ret;
+
```
missing a cJSON_Delete

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -88,19 +298,148 @@ void json_get_node_info_serialize_request(const serializer_t* const s,
+  if (cJSON_IsNumber(json_value)) {
+    out->transactionsToRequest = json_value->valueint;
+  }
+
```
cJSON_Delete

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -69,9 +226,62 @@ size_t json_get_inclusion_state_deserialize_response_get_size(
+    cJSON_Delete(json_obj);
+    return RC_CCLIENT_JSON_PARSE;
+  }
+
```
cJSON_Delete

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -52,9 +188,30 @@ void json_get_inclusion_state_serialize_request(
+  if (ret != RC_OK) {
+    return ret;
+  }
+
```
cJSON_Delete

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -33,9 +142,36 @@ void json_get_balances_serialize_request(const serializer_t* const s,
+      json_array_to_flex_trirts_array(json_obj, "references", &out->references,
+                                      NUM_TRITS_ADDRESS, NUM_TRYTES_ADDRESS);
+  if (ret != RC_OK) return ret;
+
```
cJSON_Delete

th0br0 commented on cclient/serialization/json/json_serializer.c:
```diff
@@ -8,13 +8,122 @@ void init_json_serializer(serializer_t* serializer) {
+  retcode_t ret =
+      json_array_to_flex_trirts_array(json_obj, "hashes", &out->transactions,
+                                      NUM_TRITS_ADDRESS, NUM_TRYTES_ADDRESS);
+  if (ret != RC_OK) return ret;
```
cJSON_Delete

th0br0 reviewed (changes requested):
Looking pretty good, mainly just a memleak that needs fixing

oopsmonk commented:
not merge, will do a new implementation in another PR.
